baseband/guppi/payload.py

```diff
@@ -0,0 +1,239 @@
+        # (nsample, nchan, npol)
+        if time_ordered:
+            data = data.transpose(2, 0, 1)
+        else:
+            data = data.transpose(0, 2, 1)
+        if complex_data:
+            data = data.view((data.real.dtype, (2,)))
+        words = encoder(data).ravel().view(cls._dtype_word)
+        return cls(words, sample_shape=sample_shape, bps=bps,
+                   complex_data=complex_data, time_ordered=time_ordered)
+
+    def __len__(self):
+        """Number of samples in the payload."""
+        return self.nbytes * 8 // self._bpfs // (
+            self.sample_shape.nchan if self.time_ordered else 1)
+
+    def __getitem__(self, item=()):
+        #  GUPPI data may be stored as (nsample, nchan, npol) or, if
+        # "time-ordered", (nchan, nsample, npol), both of which require
+        # reshaping to get the usual order of (nsample, npol, nchan).
+        decoder = self._decoders[self._coder]
+
+        # If we want to decode the entire dataset.
+        if item is () or item == slice(None):
+            data = decoder(self.words)
+            if self.complex_data:
+                data = data.view(self.dtype)
+            if self.time_ordered:
+                # Reshape to (nchan, nsample, npol); transpose to usual order.
+                return (data.reshape(self.sample_shape[1], -1,
+                                     self.sample_shape[0])
+                        .transpose(1, 2, 0))
+            else:
+                # Reshape to (nsample, nchan, npol); transpose to usual order.
+                return (data.reshape(-1, self.sample_shape[1],
+                                     self.sample_shape[0])
+                        .transpose(0, 2, 1))
+
+        words_slice, data_slice = self._item_to_slices(item)
+
+        if self.time_ordered:
+            # Reshape words so channels fall along first axis, then decode.
+            decoded_words = decoder(
+                self.words.reshape(self.sample_shape[1], -1)[:, words_slice])
+            return (decoded_words.view(self.dtype).T
+                    .reshape(-1, *self.sample_shape)[data_slice])
+        else:
+            # data_slice assumes (npol, nchan), so transpose before using it.
+            return (decoder(self.words[words_slice]).view(self.dtype)
+                    .reshape(-1, self.sample_shape[1], self.sample_shape[0])
+                    .transpose(0, 2, 1)[data_slice])
+
+    def __setitem__(self, item, data):
+        if item is () or item == slice(None):
+            words_slice = data_slice = slice(None)
+        else:
+            words_slice, data_slice = self._item_to_slices(item)
+
+        data = np.asanyarray(data)
+        # Check if the new data spans an entire word and is correctly shaped.
+        # If so, skip decoding.  If not, decode appropriate words and insert
+        # new data.
+        if not (data_slice == slice(None) and
+                data.shape[-len(self.sample_shape):] == self.sample_shape and
```

Comment:
Just `data.shape[-2:] == ...`, since here we know this is true.

Reply:
Changed.